example/src/App.js

```diff
@@ -1,78 +1,85 @@
+import { useRouting, Link, LocationProvider, useRoutes } from '@kaliber/routing'
+import { routeMap } from './routeMap'
 
 export default function App({ initialLocation }) {
   return (
-    <LocationProvider {...{ initialLocation }}>
+    <LocationProvider {...{ initialLocation, routeMap }}>
       <Page />
     </LocationProvider>
   )
 }
 
 function Page() {
   const { routes } = useRouting()
+  const { home, articles, notFound } = useRoutes()
 
   return (
     <>
       <Navigation />
       {routes(
-        ['', <Home />],
-        ['articles', <Articles />],
-        ['articles/:articleId/*', params => <Article {...{ params }} />],
-        ['*', params => <NotFound {...{ params }} />],
+        [home, <Home />], // eslint-disable-line react/jsx-key
+        [articles, <Articles />], // eslint-disable-line react/jsx-key
+        [articles.article, params => <Article {...{ params }} />],
+        [notFound, params => <NotFound {...{ params }} />],
       )}
     </>
   )
 }
 
 function Navigation() {
+  const { home, articles } = useRoutes()
   return (
     <div>
-      <Link to={''}>Home</Link>
-      <Link to={'articles'}>Articles</Link>
-      <Link to={interpolate('articles/:article/*', { article: 'article1' })}>Featured article</Link>
+      <Link to={home()}>Home</Link>
+      <Link to={articles()}>Articles</Link>
+      <Link to={articles.article({ id: 'article1' })}>Featured article</Link>
     </div>
   )
 }
 
 function Home() {
+  const { articles } = useRoutes()
+
   return (
     <div>
       Home
-      <Link to='articles'>Articles</Link>
+      <Link to={articles()}>Articles</Link>
     </div>
   )
 }
 
 function Articles() {
+  const { article } = useRoutes()
   return (
     <div>
       articles
       <div>
-        <Link to='article1'>article 1</Link><br />
-        <Link to='article2'>article 2</Link>
+        <Link to={article({ id: 'article1' })}>article 1</Link><br />
+        <Link to={article({ id: 'article2' })}>article 2</Link>
       </div>
     </div>
   )
 }
 
-function Article({ params: { articleId } }) {
+function Article({ params: { id } }) {
   const { routes, route } = useRouting()
+  const { main, tab1, tab2 } = useRoutes()
   return (
     <div>
-      <h1>Article {articleId}</h1>
+      <h1>Article {id}</h1>
       <div>
-        <Link to=''>Main</Link>
-        <Link to='tab1'>Tab1</Link>
-        <Link to='tab2'>Tab2</Link>
+        <Link to={main()}>Main</Link>
+        <Link to={tab1()}>Tab1</Link>
+        <Link to={tab2()}>Tab2</Link>
       </div>
       <div>
         {routes(
-          ['', 'Main content'],
-          ['tab1', 'Tab 1'],
-          ['tab2', 'Tab 2'],
-        )}
+          [main, 'Main content'],
+          [tab1, 'Tab 1'],
+          [tab2, 'Tab 2'],
+        ) || 'not matched'}
       </div>
-      {route('tab1', <div>Side bar for tab 1</div>)}
+      {route(tab1, <div>Side bar for tab 1</div>)}
```

Comment:
Los van dit PR: volgensmij zou je hier altijd een functie willen. Als je hier een pagina hebt staan, wordt hier de virtual DOM al van opgebouwd, terwijl dat pas nodig is als de route matched.

Reply:
Niet helemaal. Alleen datgene wat je hier ziet. Er wordt alleen `React.createElement` aangeroepen. Als dat een custom component is geef je alleen de functie mee, maar die wordt nog niet uitgevoerd.

Voor zover ik weet wordt de rest pas uitgevoerd bij een daadwerkelijke render.

Reply:
In dit voorbeeld komt de `console.log` niet voorbij:
```js
function App() {
  const x = <Test />

  return 'test'
}

function Test() {
  console.log('render Test')
  return null
}
```

Reply:
Ah, right! Dan heb ik niks gezegd